builder/builder.go

```diff
@@ -0,0 +1,77 @@
+package builder
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/docker/docker/daemon"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/runconfig"
+)
+
+type ImageID string
+
+type Builder interface {
+	// TODO: make this return a reference and remove ImageID
+	Build(Docker, Context) ImageID
+}
+
+type ImageCache interface {
+	GetCachedImage(ImageID, *runconfig.Config) (ImageID, error)
+	Cache(ImageID, *runconfig.Config) error
+}
+
+type Docker interface {
+	// use digest reference instead of name
+	LookupImage(name string) (*image.Image, error)
+	Pull(name string) (*image.Image, error)
+
+	// move daemon.Container to its own package
+	Container(id string) (*daemon.Container, error)
+	Create(*runconfig.Config, *runconfig.HostConfig) (*daemon.Container, error)
+	Remove(id string, cfg *daemon.ContainerRmConfig) error
+	Commit(*daemon.Container, *daemon.ContainerCommitConfig) (*image.Image, error)
+	Copy(c *daemon.Container, destPath string, srcInfo ContextEntry, decompress bool) error
+}
+
+type Context interface {
+	io.Closer
+	Get(path string) (ContextEntry, error)
```

Comment:
Yeah was thinking about this too, feel free to give more input and I'll update accordingly.